mycelo/genesis/genesis_state.go

```diff
@@ -608,87 +676,303 @@ func (ctx *deployContext) deployReserve() error {
+		logger.Info("Register group")
+		if err := validators.SimpleCallFrom(address, "registerValidatorGroup", groupComission); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (ctx *deployContext) addValidatorsToGroups() error {
+	validators := ctx.contract("Validators")
+
+	validatorGroups := ctx.accounts.ValidatorGroups()
+	for groupIdx, group := range validatorGroups {
+		groupAddress := group.Address
+		prevGroupAddress := common.ZeroAddress
+		if groupIdx > 0 {
+			prevGroupAddress = validatorGroups[groupIdx-1].Address
+		}
+
+		for i, validator := range group.Validators {
+			ctx.logger.Info("Add validator to group", "validator", validator.Address, "group", groupAddress)
+
+			// affiliate validators to group
+			if err := validators.SimpleCallFrom(validator.Address, "affiliate", groupAddress); err != nil {
 				return err
 			}
+
+			// accept validator as group member
+			if i == 0 {
+				if err := validators.SimpleCallFrom(groupAddress, "addFirstMember", validator.Address, common.ZeroAddress, prevGroupAddress); err != nil {
+					return err
+				}
+			} else {
+				if err := validators.SimpleCallFrom(groupAddress, "addMember", validator.Address); err != nil {
+					return err
+				}
+			}
 		}
+	}
 
-		ctx.logger.Info("Reporting price of StableToken to oracle")
-		err = ctx.contract("SortedOracles").SimpleCall("report",
-			stableTokenAddress,
-			ctx.GenesisConfig.StableToken.GoldPrice.BigInt(),
-			common.ZeroAddress,
-			common.ZeroAddress,
-		)
-		if err != nil {
+	return nil
+}
+
+func (ctx *deployContext) voteForGroups() error {
+	election := ctx.contract("Election")
+
+	validatorGroups := ctx.accounts.ValidatorGroupAccounts()
+
+	// value previously locked on registerValidatorGroups()
+	lockedGoldOnGroup := new(big.Int).Mul(
+		ctx.genesisConfig.Validators.GroupLockedGoldRequirements.Value,
+		big.NewInt(int64(ctx.accounts.ValidatorsPerGroup)),
+	)
+
+	// current group order (see `addFirstMember` on addValidatorsToGroup) is:
+	// [ 1stGroup, 2ndGroup, ..., lastgroup]
+
+	// each group votes for themselves.
+	// each group votes the SAME AMOUNT
+	// each group starts with 0 votes
+
+	// so, everytime we vote, that group becomes the one with most votes (most or equal)
+	// hence, we use:
+	//    greater = zero (we become the one with most votes)
+	//    lesser = currentLeader
+
+	// special case: only one group (no lesser or greater)
+	if len(validatorGroups) == 1 {
+		groupAddress := validatorGroups[0].Address
+		ctx.logger.Info("Vote for group", "group", groupAddress, "amount", lockedGoldOnGroup)
+		if err := election.SimpleCallFrom(groupAddress, "vote", groupAddress, lockedGoldOnGroup, common.ZeroAddress, common.ZeroAddress); err != nil {
 			return err
 		}
+	}
 
-		ctx.logger.Info("Add StableToken to the reserve")
-		err = ctx.contract("Reserve").SimpleCall("addToken", stableTokenAddress)
-		if err != nil {
+	// first to vote is group 1, which is already the leader. Hence lesser should go to group2
```

Comment:
```suggestion
	// first to vote is group 0, which is already the leader. Hence lesser should go to group 1
```

The whole voting mechanism is pretty complex due to lesser and great thing.  But I think the comments here were also confusing me to their use of 1-based indexing whereas in the code indexing is 0-based.

Reply:
This also raises the question of whether we should make the validators and validator group names start at zero, but I don't think we really have to because of this.  At least in the context of this function's comments I think referring to them by their index in the array makes more sense than by the 1-based index in their names.

Reply:
Changed to 0 based index